### PR TITLE
Add total ring of fractions

### PR DESCRIPTION
**Pull request by wbhart**
This also adds is_zero_divisor (required by total ring of fractions) for the following in AbstractAlgebra:

* PolyElem
* LaurentPoly
* MPoly
* FieldElem
* Rational
* Integer
* ResElem
* UnivPolyElem

I would be obliged if the reviewer could check these functions carefully to ensure I defined them correctly.

I have not added this function for the following:

* Absolute and Relative Series (including Laurent and Puiseux) as divisibility doesn't make much sense in inexact rings
* Matrix Algebras (I don't know an algorithm)
* NCRings (because I got tired and the total ring of fractions won't work there anyway)

### CODE REVIEW

**Comment by wbhart** [comment]
I think this is ready for review now.

File: src/generic/TotalFraction.jl [hunk 1]
```diff
@@ -0,0 +1,613 @@
...
+end
+
+function (a::TotFracRing{T})(b::TotFrac{T}) where {T <: RingElement}
+   a != parent(b) && error("Could not coerce to fraction")
+   return b
+end
+
+###############################################################################
+#
+#   FractionField constructor
```
Comment (tthsqe12) [inline src/generic/TotalFraction.jl:594]: TotalRingOfFractions

File: src/generic/TotalFraction.jl [hunk 2]
```diff
@@ -0,0 +1,613 @@
...
+###############################################################################
+
+@doc Markdown.doc"""
+    Base.inv(a::TotFrac)
+
+Return the inverse of the fraction $a$ if it exists, otherwise raise an
+exception.
+"""
+function Base.inv(a::TotFrac)
+   is_zero_divisor(numerator(a, false)) && throw(DivideError())
```
Comment (tthsqe12) [inline src/generic/TotalFraction.jl:353]: `throw(NotInvertibleError(a))` is more apt

File: src/generic/TotalFraction.jl [hunk 3]
```diff
@@ -0,0 +1,613 @@
...
+function Base.numerator(a::TotFrac, canonicalise::Bool=true)
+   return a.num
+end
+
+function Base.denominator(a::TotFrac, canonicalise::Bool=true)
+   return a.den
+end
+
+function deepcopy_internal(a::TotFrac{T}, dict::IdDict) where {T <: RingElem}
+   v = TotFrac{T}(deepcopy(numerator(a, false)), deepcopy(denominator(a, false)))
```
Comment (tthsqe12) [inline src/generic/TotalFraction.jl:86]: I think these deepcopys should pass the iddict along

File: test/generic/TotalFraction-test.jl [hunk 4]
```diff
@@ -0,0 +1,287 @@
+@testset "Generic.Frac.constructors" begin
```
Comment (tthsqe12) [inline test/generic/TotalFraction-test.jl:1]: all of these `Frac`s in these test set names should be `TotFrac`s

**Review by tthsqe12 — CHANGES_REQUESTED** [review]

File: src/generic/TotalFraction.jl (hunk 1 above)
Comment (wbhart) [reply src/generic/TotalFraction.jl:594]: Thanks. Fixed.

File: src/generic/TotalFraction.jl (hunk 3 above)
Comment (wbhart) [reply src/generic/TotalFraction.jl:86]: Ok. I also fixed the same for FractionField as it was done the same way there.

File: test/generic/TotalFraction-test.jl (hunk 4 above)
Comment (wbhart) [reply test/generic/TotalFraction-test.jl:1]: Thanks. Fixed.

File: src/generic/TotalFraction.jl (hunk 3 above)
Comment (tthsqe12) [reply src/generic/TotalFraction.jl]: pass along via another deepcopy_internal

Comment (wbhart) [reply src/generic/TotalFraction.jl]: ok, fixed.

**Comment by tthsqe12** [comment]
error is from throwing a NotInvertibleError instead of integer division error

**Comment by wbhart** [comment]
Let's see if it passes this time.

**Pull request merged**